File: sdk/src/main/java/com/cxense/cxensesdk/ConversionEventConverter.java
package com.cxense.cxensesdk;

import androidx.annotation.NonNull;

import com.cxense.cxensesdk.db.EventRecord;
import com.cxense.cxensesdk.model.ConversionEvent;
import com.cxense.cxensesdk.model.Event;
import com.google.gson.Gson;

import java.util.Map;

class ConversionEventConverter extends EventConverter<ConversionEvent> {
    private final Gson gson;

    ConversionEventConverter(Gson gson) {
        this.gson = gson;
    }

    @Override
    public boolean canConvert(@NonNull Event event) {
        return event instanceof ConversionEvent;
    }

    @Override
    public Map<String, String> toQueryMap(@NonNull ConversionEvent event) {
        return null;
    }

    @NonNull
    @Override
    public EventRecord toEventRecord(@NonNull ConversionEvent event) {
        EventRecord record = new EventRecord();
        record.customId = event.getEventId();
        record.data = gson.toJson(event);
        record.timestamp = System.currentTimeMillis();
        record.eventType = event.getType();
        return record;
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/ConversionEventConverter.kt
package com.cxense.cxensesdk

import androidx.annotation.RestrictTo
import com.cxense.cxensesdk.db.EventRecord
import com.cxense.cxensesdk.model.ConversionEvent
import com.cxense.cxensesdk.model.Event
import com.squareup.moshi.JsonAdapter

/**
 * Supports [ConversionEvent] to [EventRecord] converting
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
class ConversionEventConverter(
    private val jsonAdapter: JsonAdapter<ConversionEvent>
) : EventConverter() {
    override fun canConvert(event: Event): Boolean = event is ConversionEvent

    override fun toEventRecord(event: Event): EventRecord? =
        (event as? ConversionEvent)?.run {
            EventRecord(
                ConversionEvent.EVENT_TYPE,
                eventId,
                jsonAdapter.toJson(this),
                mergeKey = mergeKey
            )
        }

    override fun update(oldRecord: EventRecord, event: Event): EventRecord =
        with(event as ConversionEvent) {
            jsonAdapter.fromJson(oldRecord.data)?.let { old ->
                toEventRecord(
                    ConversionEvent(
                        identities.takeUnless { it.isEmpty() } ?: old.identities,
                        siteId,
                        consentOptions.takeUnless { it.isEmpty() } ?: old.consentOptions,
                        productId,
                        funnelStep,
                        price ?: old.price,
                        renewalFrequency ?: old.renewalFrequency,
                        eventType
                    )
                )
            } ?: oldRecord
        }
}
